File: python/examples/auto_chunking.py
import asyncio
import time
from collections.abc import AsyncGenerator
from typing import Any

import uvloop

from camera_ui_rpc import RPCClass, create_rpc_client

small_data = "This is a small response that fits in a single message"
medium_data = bytearray(5 * 1024 * 1024)  # 5MB
large_data = bytearray(50 * 1024 * 1024)  # 50MB

for i in range(len(medium_data)):
    medium_data[i] = i % 256

for i in range(len(large_data)):
    large_data[i] = i % 256

chunk_buffers: list[bytes] = []

for i in range(3):
    buffer = bytearray(5 * 1024 * 1024)
    for j in range(len(buffer)):
        buffer[j] = (i + j) % 256
    chunk_buffers.append(bytes(buffer))

echo_data = bytearray(4 * 1024 * 1024)
for i in range(len(echo_data)):
    echo_data[i] = (i * 2) % 256
echo_data = bytes(echo_data)


@RPCClass
class TestService:
    async def get_small_data(self) -> str:
        print(f"Returning small data: {small_data}")
        return small_data

    async def get_medium_data(self) -> bytes:
        print("Returning 5MB buffer")
        return bytes(medium_data)

    async def get_large_data(self) -> bytes:
        print("Returning 50MB buffer")
        return bytes(large_data)

    async def echo(self, data: bytes) -> bytes:
        print(f"Echoing {len(data) / 1024 / 1024:.2f}MB")
        return data

    async def generate_large_data_stream(self) -> AsyncGenerator[bytes, None]:
        print("Streaming 3 chunks of 5MB each")

        for i in range(3):
            yield chunk_buffers[i]
            await asyncio.sleep(0)

    async def ping(self) -> str:
        return f"pong at {time.strftime('%Y-%m-%d %H:%M:%S')}"


@RPCClass
class SmallPayloadService:
    async def get_small_data(self) -> str:
        return "Small response that fits in 1KB"

    async def get_medium_data(self) -> bytes:
        return bytes(10 * 1024)

    async def get_large_data(self) -> bytes:
        return bytes(5 * 1024 * 1024)


async def main():
    print("Automatic Chunking Test\n")
    print("Testing transparent chunking in RPC library")
    print("Note: NATS server default max_payload is typically 1MB\n")

    server = create_rpc_client(
        {
            "servers": ["nats://localhost:4222"],
            "auth": {"user": "server", "password": "server_password"},
            "name": "test-service",
        }
    )

    await server.connect()
    unsub_test = await server.register_handler("test", TestService())

    client = create_rpc_client(
        {
            "servers": ["nats://localhost:4222"],
            "auth": {"user": "server", "password": "server_password"},
            "name": "test-client",
        }
    )

    await client.connect()
    service = client.create_proxy("test", TestService)

    total_start = time.perf_counter()

    print("Small Data")
    small_start = time.perf_counter()
    small_data = await service.get_small_data()
    print(f'Small data received in {(time.perf_counter() - small_start) * 1000:.0f}ms: "{small_data}"')

    print("\nMedium Data (5MB)")
    medium_start = time.perf_counter()
    medium_data = await service.get_medium_data()
    medium_time = time.perf_counter() - medium_start
    print(f"Medium data received: {len(medium_data) / 1024 / 1024:.2f}MB in {medium_time * 1000:.0f}ms")
    print(f"Throughput: {5 / medium_time:.2f} MB/s")

    corrupted = False
    for i in range(min(1000, len(medium_data))):
        if medium_data[i] != i % 256:
            corrupted = True
            break
    print(f"Data integrity: {'CORRUPTED' if corrupted else 'OK'}")

    print("\nLarge Data (50MB)")
    large_start = time.perf_counter()
    large_data = await service.get_large_data()
    large_time = time.perf_counter() - large_start
    print(f"Large data received: {len(large_data) / 1024 / 1024:.2f}MB in {large_time * 1000:.0f}ms")
    print(f"Throughput: {50 / large_time:.2f} MB/s")

    print("\nEcho Test (4MB round-trip)")
    echo_start = time.perf_counter()
    echoed = await service.echo(echo_data)
    echo_time = time.perf_counter() - echo_start
    print(f"Echo completed in {echo_time * 1000:.0f}ms")
    print(f"Round-trip throughput: {8 / echo_time:.2f} MB/s")
    print(f"Data matches: {'yes' if echo_data == echoed else 'no'}")

    print("\nStreaming Large Chunks")
    stream_start = time.perf_counter()
    streamed_bytes = 0
    chunk_count = 0

    async for chunk in service.generate_large_data_stream():
        chunk_count += 1
        chunk_size = len(chunk) if chunk else 0
        streamed_bytes += chunk_size
        print(f"Received stream chunk {chunk_count}: {chunk_size / 1024 / 1024:.2f}MB")

    stream_time = time.perf_counter() - stream_start
    print(f"Stream completed: {streamed_bytes / 1024 / 1024:.2f}MB in {stream_time * 1000:.0f}ms")
    if streamed_bytes == 0:
        print("Known issue: Buffer streaming through MessagePack may lose data")
    else:
        print(f"Stream throughput: {streamed_bytes / 1024 / 1024 / stream_time:.2f} MB/s")

    print("\nConcurrent Operations")
    concurrent_tasks: list[asyncio.Task[Any]] = []

    async def large_transfer():
        data = await service.get_large_data()
        print(f"Large data received: {len(data) / 1024 / 1024:.2f}MB")

    concurrent_tasks.append(asyncio.create_task(large_transfer()))

    for i in range(5):

        async def ping_test(index: int):
            await asyncio.sleep(index * 0.1)
            ping_start = time.perf_counter()
            await service.ping()
            print(f"Ping {index + 1} latency: {(time.perf_counter() - ping_start) * 1000:.0f}ms")

        concurrent_tasks.append(asyncio.create_task(ping_test(i)))

    await asyncio.gather(*concurrent_tasks)

    print("\nExtreme Chunking (1KB max payload)")

    small_payload_server = create_rpc_client(
        {
            "servers": ["nats://localhost:4222"],
            "auth": {"user": "server", "password": "server_password"},
            "name": "small-payload-server",
            "max_payload_size": 1024,
        }
    )

    await small_payload_server.connect()

    unsub_small = await small_payload_server.register_handler("test-small-payload", SmallPayloadService())

    small_payload_client = create_rpc_client(
        {
            "servers": ["nats://localhost:4222"],
            "auth": {"user": "server", "password": "server_password"},
            "name": "small-payload-client",
            "max_payload_size": 1024,
        }
    )

    await small_payload_client.connect()
    small_payload_service = small_payload_client.create_proxy("test-small-payload", SmallPayloadService)

    small_result = await small_payload_service.get_small_data()
    print(f'Small data with 1KB limit: "{small_result}"')

    start = time.perf_counter()
    medium_buffer = await small_payload_service.get_medium_data()
    elapsed = time.perf_counter() - start
    expected_chunks = (10 * 1024 + 1023) // 1024
    print(
        f"10KB data with 1KB limit: received {len(medium_buffer)} bytes in {elapsed * 1000:.0f}ms (~{expected_chunks} chunks)"
    )

    large_start_new = time.perf_counter()
    large_buffer = await small_payload_service.get_large_data()
    large_elapsed = time.perf_counter() - large_start_new
    expected_chunks = (5 * 1024 * 1024 + 1023) // 1024
    print(
        f"5MB data with 1KB limit: received {len(large_buffer)} bytes in {large_elapsed * 1000:.0f}ms (~{expected_chunks} chunks)"
    )

    print("\nCleaning up...")
    await unsub_small()
    await small_payload_client.disconnect()
    await small_payload_server.disconnect()

    await unsub_test()
    await client.disconnect()
    await server.disconnect()

    total_elapsed = (time.perf_counter() - total_start) * 1000
    print(f"\nAll tests completed! Total time: {total_elapsed:.1f}ms")


if __name__ == "__main__":
    try:
        uvloop.run(main())
    except KeyboardInterrupt:
        print("\nTest interrupted by user")
    except Exception as e:
        print(f"Error in native channel request test: {e}")
        import traceback

        traceback.print_exc()

File: node/examples/auto-chunking.ts
import { RPCClass } from '../src/decorators.js';
import { createRPCClient } from '../src/index.js';

const smallData = 'This is a small response that fits in a single message';
const mediumData = Buffer.alloc(5 * 1024 * 1024); // 5MB
const largeData = Buffer.alloc(50 * 1024 * 1024); // 50MB

for (let i = 0; i < mediumData.length; i++) {
  mediumData[i] = i % 256;
}

for (let i = 0; i < largeData.length; i++) {
  largeData[i] = i % 256;
}

const chunkBuffers: Buffer[] = [];
for (let i = 0; i < 3; i++) {
  const buffer = Buffer.alloc(5 * 1024 * 1024);
  for (let j = 0; j < buffer.length; j++) {
    buffer[j] = (i + j) % 256;
  }
  chunkBuffers.push(buffer);
}

const echoData = Buffer.alloc(4 * 1024 * 1024);
for (let i = 0; i < echoData.length; i++) {
  echoData[i] = (i * 2) % 256;
}

interface TestService {
  getSmallData(): Promise<string>;
  getMediumData(): Promise<Buffer>;
  getLargeData(): Promise<Buffer>;

  echo(data: Buffer): Promise<Buffer>;

  generateLargeDataStream(): AsyncGenerator<Buffer>;

  ping(): Promise<string>;
}

@RPCClass
class TestServiceImpl implements TestService {
  async getSmallData() {
    return smallData;
  }

  async getMediumData() {
    console.log('Returning 5MB buffer');
    return mediumData;
  }

  async getLargeData() {
    console.log('Returning 50MB buffer');
    return largeData;
  }

  async echo(data: Buffer) {
    console.log(`Echoing ${(data.length / 1024 / 1024).toFixed(2)}MB`);
    return data;
  }

  async *generateLargeDataStream() {
    console.log('Streaming 3 chunks of 5MB each');

    for (let i = 0; i < 3; i++) {
      console.log(`Yielding chunk ${i + 1}/3 - size: ${chunkBuffers[i].length} bytes`);
      console.log(`First bytes: [${Array.from(chunkBuffers[i].slice(0, 10)).join(', ')}]`);
      yield chunkBuffers[i];

      await new Promise((resolve) => setImmediate(resolve));
    }
  }

  async ping() {
    return `pong at ${new Date().toISOString()}`;
  }
}

async function main() {
  const totalStart = performance.now();

  console.log('Automatic Chunking Test\n');
  console.log('Testing transparent chunking in RPC library');
  console.log('Note: NATS server default max_payload is typically 1MB');

  const client = createRPCClient({
    servers: ['nats://localhost:4222'],
    name: 'test-client',
    auth: { user: 'server', password: 'server_password' },
  });

  const server = createRPCClient({
    servers: ['nats://localhost:4222'],
    name: 'test-server',
    auth: { user: 'server', password: 'server_password' },
  });

  await client.connect();
  await server.connect();

  const unsub1 = await server.registerHandler('test', new TestServiceImpl());
  const unsub2 = await server.registerHandler('testSmall', new TestServiceImpl());

  const service = client.createProxy<TestService>('test');

  console.log('\n--- Test 1: Small Data ---');
  const smallStart = performance.now();
  const smallResult = await service.getSmallData();
  console.log(`Small data received in ${(performance.now() - smallStart).toFixed(0)}ms: "${smallResult}"`);

  console.log('\n--- Test 2: Medium Data (5MB) ---');
  const mediumStart = performance.now();
  const mediumDataResult = await service.getMediumData();
  const mediumTime = performance.now() - mediumStart;
  console.log(`Medium data received: ${(mediumDataResult.length / 1024 / 1024).toFixed(2)}MB in ${mediumTime.toFixed(0)}ms`);
  console.log(`Throughput: ${(5 / (mediumTime / 1000)).toFixed(2)} MB/s`);

  let corrupted = false;
  for (let i = 0; i < Math.min(1000, mediumDataResult.length); i++) {
    if (mediumDataResult[i] !== i % 256) {
      corrupted = true;
      break;
    }
  }
  console.log(`Data integrity: ${corrupted ? 'CORRUPTED' : 'OK'}`);

  console.log('\n--- Test 3: Large Data (50MB) ---');
  const largeStart = performance.now();
  const largeDataResult = await service.getLargeData();
  const largeTime = performance.now() - largeStart;
  console.log(`Large data received: ${(largeDataResult.length / 1024 / 1024).toFixed(2)}MB in ${largeTime.toFixed(0)}ms`);
  console.log(`Throughput: ${(50 / (largeTime / 1000)).toFixed(2)} MB/s`);

  console.log('\n--- Test 4: Echo Test (4MB round-trip) ---');
  const echoStart = performance.now();
  const echoed = await service.echo(echoData);
  const echoTime = performance.now() - echoStart;
  console.log(`Echo completed in ${echoTime.toFixed(0)}ms`);
  console.log(`Round-trip throughput: ${(8 / (echoTime / 1000)).toFixed(2)} MB/s`);
  console.log(`Data matches: ${Buffer.compare(echoData, echoed) === 0 ? 'yes' : 'no'}`);

  console.log('\n--- Test 5: Streaming Large Chunks ---');
  const streamStart = performance.now();
  let streamedBytes = 0;
  let chunkCount = 0;

  for await (const chunk of service.generateLargeDataStream()) {
    chunkCount++;
    streamedBytes += chunk.length;
    console.log(`Received stream chunk ${chunkCount}: ${(chunk.length / 1024 / 1024).toFixed(2)}MB`);
  }

  const streamTime = performance.now() - streamStart;
  console.log(`Stream completed: ${(streamedBytes / 1024 / 1024).toFixed(2)}MB in ${streamTime.toFixed(0)}ms`);
  console.log(`Stream throughput: ${(streamedBytes / 1024 / 1024 / (streamTime / 1000)).toFixed(2)} MB/s`);

  console.log('\n--- Test 6: Concurrent Operations ---');
  const concurrentTasks: Promise<void>[] = [];

  const largeTransfer = async () => {
    const data = await service.getLargeData();
    console.log(`Large data received: ${(data.length / 1024 / 1024).toFixed(2)}MB`);
  };

  concurrentTasks.push(largeTransfer());

  for (let i = 0; i < 5; i++) {
    const pingTest = async (index: number) => {
      const start = performance.now();
      await service.ping();
      console.log(`Ping ${index + 1} latency: ${(performance.now() - start).toFixed(0)}ms`);
    };
    concurrentTasks.push(pingTest(i));

    await new Promise((resolve) => setTimeout(resolve, i * 100));
  }

  await Promise.all(concurrentTasks);

  console.log('\n--- Test 7: Extreme Chunking (1KB max payload) ---');

  const clientSmall = createRPCClient({
    servers: ['nats://localhost:4222'],
    name: 'test-client-small',
    auth: { user: 'server', password: 'server_password' },
  });

  await clientSmall.connect();

  // Override max payload size to test extreme chunking
  // @ts-ignore - accessing private property for testing
  clientSmall._maxPayloadSize = 1024; // 1KB

  const serviceSmall = clientSmall.createProxy<TestService>('testSmall');

  const smallDataResult = await serviceSmall.getSmallData();
  console.log(`Small data with 1KB limit: "${smallDataResult}"`);

  const tenKB = await serviceSmall.echo(Buffer.alloc(10 * 1024, 0x42));
  console.log(`10KB data with 1KB limit: received ${tenKB.length} bytes in ${(performance.now() - streamStart).toFixed(0)}ms (~10 chunks)`);

  const fiveMBStart = performance.now();
  const fiveMBData = await serviceSmall.getMediumData();
  console.log(`5MB data with 1KB limit: received ${fiveMBData.length} bytes in ${(performance.now() - fiveMBStart).toFixed(0)}ms (~5120 chunks)`);

  await clientSmall.disconnect();

  await unsub1();
  await unsub2();
  await client.disconnect();
  await server.disconnect();

  const totalElapsed = performance.now() - totalStart;
  console.log(`\nAll tests completed! Total time: ${totalElapsed.toFixed(1)}ms`);
}

main().catch(console.error);
